File: 프로그래머스/2/87946. 피로도/피로도.java
class Solution {
    public int solution(int k, int[][] dungeons) {
        
        boolean[] visited = new boolean[dungeons.length];
        return dfs(k, dungeons, visited);
    }
    
    public int dfs(int k, int[][] dungeons, boolean[] visited) {
        int answer = 0;
        
        for (int i = 0; i < dungeons.length; i++) {
            if (!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                answer = Math.max(answer, Math.max(dfs(k, dungeons, visited), 1 + dfs(k - dungeons[i][1], dungeons, visited)));
                visited[i] = false;
            }
        }

        
        return answer;
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
from itertools import permutations

def solution(k, dungeons):

    answer = 0
    for per in permutations(dungeons, len(dungeons)):
        result = 0
        piro = k
        for min_piro, somo_piro in per:
            if piro >= min_piro:
                result += 1
                piro -= somo_piro
        answer = max(answer, result)
              
    return answer